source/extensions/filters/http/oauth/filter.cc

```diff
@@ -0,0 +1,475 @@
+const std::string& defaultOauthSignout() { CONSTRUCT_ON_FIRST_USE(std::string, "/_signout"); }
+
+const std::string& queryParamsError() { CONSTRUCT_ON_FIRST_USE(std::string, "error"); }
+const std::string& queryParamsCode() { CONSTRUCT_ON_FIRST_USE(std::string, "code"); }
+const std::string& queryParamsState() { CONSTRUCT_ON_FIRST_USE(std::string, "state"); }
+
+FilterConfig::FilterConfig(
+    const envoy::extensions::filters::http::oauth::v3::OAuth2Config& proto_config,
+    Upstream::ClusterManager& cluster_manager, std::shared_ptr<SecretReader> secret_reader,
+    Stats::Scope& scope, const std::string& stats_prefix)
+    : cluster_name_(proto_config.cluster()), client_id_(proto_config.credentials().client_id()),
+      oauth_server_hostname_(proto_config.hostname()),
+      callback_path_(
+          PROTOBUF_GET_STRING_OR_DEFAULT(proto_config, callback_path, defaultOauthCallback())),
+      signout_path_(
+          PROTOBUF_GET_STRING_OR_DEFAULT(proto_config, signout_path, defaultOauthSignout())),
+      forward_bearer_token_(proto_config.forward_bearer_token()),
+      pass_through_options_method_(proto_config.pass_through_options_method()),
+      secret_reader_(secret_reader), stats_(FilterConfig::generateStats(stats_prefix, scope)) {
+  if (!cluster_manager.get(cluster_name_)) {
+    throw EnvoyException(fmt::format("OAuth2 filter: unknown cluster '{}' in config. Please "
+                                     "specify which cluster to direct OAuth requests to.",
+                                     cluster_name_));
+  }
+}
+
+FilterStats FilterConfig::generateStats(const std::string& prefix, Stats::Scope& scope) {
+  return {ALL_OAUTH_FILTER_STATS(POOL_COUNTER_PREFIX(scope, prefix))};
+}
+
+void OAuth2CookieValidator::setParams(const Http::RequestHeaderMap& headers,
+                                      const std::string& secret) {
+  expires_ = Http::Utility::parseCookieValue(headers, "OauthExpires");
+  username_ = Http::Utility::parseCookieValue(headers, "OauthUsername");
+  token_ = Http::Utility::parseCookieValue(headers, "BearerToken");
+  hmac_ = Http::Utility::parseCookieValue(headers, "OauthHMAC");
+  host_ = headers.Host()->value().getStringView();
+
+  std::vector<uint8_t> vec(secret.begin(), secret.end());
+  secret_ = std::move(vec);
+}
+
+bool OAuth2CookieValidator::hmacIsValid() const {
+  auto& crypto_util = Envoy::Common::Crypto::UtilitySingleton::get();
+  const std::string hmac_payload = absl::StrCat(host_, username_, expires_, token_);
+  const std::string pre_encoded_hmac =
+      Hex::encode(crypto_util.getSha256Hmac(secret_, hmac_payload));
+  std::string encoded_hmac;
+  absl::Base64Escape(pre_encoded_hmac, &encoded_hmac);
+
+  return encoded_hmac == hmac_;
+}
+
+bool OAuth2CookieValidator::timestampIsValid() const {
+  std::time_t epoch = std::time(nullptr);
+  int expires;
+  try {
+    expires = std::stoi(expires_);
+  } catch (const std::exception&) {
```

Comment:
use `absl::SimpleAtoi` to avoid dealing with exceptions. 

Also, should `expires_` be a numeric value instead, letting us do the parsing in `setParams`?